### PR TITLE
[Pallas] Add block-sparse kernel tutorial

### PR DESCRIPTION
**Pull request by justinjfu**
Adds a tutorial covering:
- Scalar prefetch
- Sparse-data kernels (COO format)
- Sparse computation kernels

### CODE REVIEW

File: docs/pallas/tpu/sparse.md [hunk 1]
```diff
@@ -0,0 +1,571 @@
...
+  def __init__(self,
+    num_scalar_prefetch: int,
+    grid: tuple[int, ...],
+    in_specs: PyTree[BlockSpec],
+    out_specs: PyTree[BlockSpec],
+    scratch_shapes: tuple[MemorySpace, ...]):
+      ...
+```
+
+There are two additional agruments for `pltpu.PrefetchScalarGridSpec` compared to `pl.GridSpec`:
```
Comment (sharadmv) [inline docs/pallas/tpu/sparse.md:67]: agruments -> arguments

File: docs/pallas/tpu/sparse.md [hunk 2]
```diff
@@ -0,0 +1,571 @@
...
+
+assert "TPU" in jax.devices()[0].device_kind, "Please run this notebook with TPU devices."
+print("Running on", jax.devices()[0].device_kind)
+```
+
++++ {"id": "FIDGpPTEIcOa"}
+
+## Dynamic Block Indexing with Scalar Prefetch
+
+We will be exploiting the "scalar prefetch" feature of Pallas to enable us to write sparse kernels. Scalar prefetch allows you to pass in a small amount of data into SMEM ("scalar") that is loaded before the invocation of the kernel ("prefetch"). This metadata is then passed into the `index_map` for each BlockSpec, allowing the you to perform data-dependent indexing calculations. The main goal of this tutorial is to go over common programming patterns that utilize this feature.
```
Comment (sharadmv) [inline docs/pallas/tpu/sparse.md:52]: SMEM ("scalar") -> SMEM ("scalar memory")

You could also include the same TPU memory space diagram here to provide some more context (like normally we fetch things into VMEM but we also have SMEM available for us to use).

Comment (sharadmv) [inline docs/pallas/tpu/sparse.md:52]: "before the invocation of the kernel" -> "before the start of the pipeline"

Comment (sharadmv) [inline docs/pallas/tpu/sparse.md:52]: "This metadata is then passed into the `index_map` for each BlockSpec" -> "Because this data is before the pipeline, it is available for use in our `index_map`s in each BlockSpec" 

or something like that

File: docs/pallas/tpu/sparse.md [hunk 3]
```diff
@@ -0,0 +1,571 @@
...
+
+To enable scalar prefetch, use the `PrefetchScalarGridSpec` in place of the standard `pl.GridSpec`:
```
Comment (sharadmv) [inline docs/pallas/tpu/sparse.md:54]: Nit: enable -> use

File: docs/pallas/tpu/sparse.md [hunk 4]
```diff
@@ -0,0 +1,571 @@
...
+def kernel(*prefetch_refs, *input_refs, *output_refs, *scratch_refs):
+    ...
+```
+
+- When calling a new kernel using `pallas_call`, the function returned by `pallas_call` also expects the scalar prefetch arguments to come before the inputs, e.g.
+```python
+kernel = pl.pallas_call(...)
+result = kernel(*prefetch_args, *input_args)
+```
+
```
Comment (sharadmv) [inline docs/pallas/tpu/sparse.md:90]: Can you add a note about Pytrees here? 

I.e. scalar prefetch values can be Pytrees. All of of their leaf values will be prefetched together and `num_scalar_prefetch` really refers to the number of arguments (i.e. pytrees) not the number of array values.

File: docs/pallas/tpu/sparse.md [hunk 5]
```diff
@@ -0,0 +1,571 @@
...
+def dynamic_slice_kernel(indices, x_ref, o_ref):
+  del indices
+  o_ref[...] = x_ref[...]
+
+@checkify.checkify
+@functools.partial(jax.jit, static_argnums=(2,))
+def block_dynamic_slice(x, starts, sizes):
+  grid_spec = pltpu.PrefetchScalarGridSpec(
+      num_scalar_prefetch=1,
+      grid=(1, 1),
```
Comment (sharadmv) [inline docs/pallas/tpu/sparse.md:127]: Can you juts use an empty grid here? `grid = ()`

File: docs/pallas/tpu/sparse.md [hunk 6]
```diff
@@ -0,0 +1,571 @@
...
+
+Of course, this kernel is limited in that our slice sizes must fit inside of a kernel block (limited by VMEM size) and we can only start on size-aligned indices. A more advanced kernel would decouple the kernel block size with the slice size and allow non-aligned start indices.
+
+```{code-cell}
+---
+executionInfo:
+  elapsed: 143
+  status: ok
+  timestamp: 1726003877561
+  user:
```
Comment (sharadmv) [inline docs/pallas/tpu/sparse.md:111]: can delete all this metadata

File: docs/pallas/tpu/sparse.md [hunk 7]
```diff
@@ -0,0 +1,571 @@
...
+err, result = block_dynamic_slice(x, starts=(128, 256), sizes=(128, 128))
+err.throw()
+ref = lax.dynamic_slice(x, start_indices=(128, 256), slice_sizes=(128, 128))
+diff = jnp.max(jnp.abs(result - ref))
+print("Error |result - lax.dynamic_slice| =", diff)
+```
+
++++ {"id": "K2dod4lkoifa"}
+
+## Sparse Kernels: Representing Sparse Data
```
Comment (sharadmv) [inline docs/pallas/tpu/sparse.md:156]: Maybe say "Block Sparse" here instead of Sparse?

File: docs/pallas/tpu/sparse.md [hunk 8]
```diff
@@ -0,0 +1,571 @@
...
+result = kernel(*args)
+
+ref = X_dense @ Y
+diff = jnp.abs(ref - result)
+print('mean |result - ref|:', jnp.mean(diff))
+```
+
++++ {"id": "2KDgPKF2tUjq"}
+
+We can do a quick benchmark to compare the performance of our sparse kernel compared to a dense matmul in JAX. On a TPU v5 lite chip, this kernel achieves a roughly ~6x speed increase compared to the theoretical 10x from the sparsity factor.
```
Comment (sharadmv) [inline docs/pallas/tpu/sparse.md:308]: v5 lite -> v5e

File: docs/pallas/tpu/sparse.md [hunk 9]
```diff
@@ -0,0 +1,571 @@
...
+print("Reference: %.3f ms (avg over %d trials)" % (time * 1000, N_trials))
+```
+
++++ {"id": "Q1KKd5vTCwnB"}
+
+## Sparse Access Patterns on Dense Data
+
+In our previous example we considered the case when the data itself is sparse. This manifested itself in the kernel structure as a dimension in the kernel grid that was dynamic and looped over the number of nonzero blocks (`num_blocks`).
+
+A second useful programming pattern emerges when the underlying is data is dense, but we wish to perform sparse computation over it. Our kernel grid in this case will be dense, but we wish to skip over some blocks in the grid as indicated by a block-sparse mask. This type of programming pattern is commonly arises when using masks in many machine learning applications, such as causal or local masks in self-attention. In these cases, we can entirely skip over computation in blocks where the mask is zeroed-out.
```
Comment (sharadmv) [inline docs/pallas/tpu/sparse.md:358]: Maybe cite/link to Splash Attention and Flex attention here?

File: docs/pallas/tpu/sparse.md [hunk 10]
```diff
@@ -0,0 +1,571 @@
...
+  user:
+    displayName: Justin Fu
+    userId: '17543197034567316452'
+  user_tz: 420
+id: MAT9JjGNvsx8
+outputId: a32d56fb-a71b-4007-c6a5-e5270dcaa6cf
+---
+N_trials = 100
+
+PallasImpl = lambda *args: kernel(*args)
```
Comment (sharadmv) [inline docs/pallas/tpu/sparse.md:564]: Nit: PallasImpl -> pallas_impl, RefImpl -> ref_impl, N_trials -> n_trials

File: docs/pallas/tpu/sparse.md (hunk 9 above)
Comment (sharadmv) [reply docs/pallas/tpu/sparse.md:358]: Oh and GMM

File: docs/pallas/tpu/sparse.md [hunk 11]
```diff
@@ -0,0 +1,571 @@
...
+result = kernel(*args)
+
+ref = mask * (X @ Y)
+diff = jnp.abs(ref - result)
+print('mean |result - ref|:', jnp.mean(diff))
+```
+
++++ {"id": "uutNGgjZGGhB"}
+
+Now let's compare performance versus a naive dense implementation. On TPU v5 lite, we achieve around a ~1.8x speed increase with the sparse kernel, compared to a theoretical best-case of 2x from using a lower triangular mask and only visiting half of the possible outputs.
```
Comment (sharadmv) [inline docs/pallas/tpu/sparse.md:543]: v5 lite -> v5e

File: docs/pallas/tpu/sparse.md (hunk 6 above)
Comment (sharadmv) [reply docs/pallas/tpu/sparse.md:111]: and elsewhere

File: docs/pallas/tpu/sparse.md (hunk 3 above)
Comment (sharadmv) [inline docs/pallas/tpu/sparse.md:54]: `PrefetchScalarGridSpec` -> `pltpu.PrefetchScalarGridSpec`

**Review by sharadmv — APPROVED** [review]
This is awesome! I love the visualization of prefetch map.

File: docs/pallas/tpu/sparse.md (hunk 1 above)
Comment (justinjfu) [reply docs/pallas/tpu/sparse.md:67]: Done.

File: docs/pallas/tpu/sparse.md (hunk 2 above)
Comment (justinjfu) [reply docs/pallas/tpu/sparse.md:52]: Done.

Comment (justinjfu) [reply docs/pallas/tpu/sparse.md:52]: Done.

Comment (justinjfu) [reply docs/pallas/tpu/sparse.md:52]: Done.

File: docs/pallas/tpu/sparse.md (hunk 3 above)
Comment (justinjfu) [reply docs/pallas/tpu/sparse.md:54]: Done.

File: docs/pallas/tpu/sparse.md (hunk 8 above)
Comment (justinjfu) [reply docs/pallas/tpu/sparse.md:308]: Done.

File: docs/pallas/tpu/sparse.md (hunk 11 above)
Comment (justinjfu) [reply docs/pallas/tpu/sparse.md:543]: Done.

File: docs/pallas/tpu/sparse.md (hunk 10 above)
Comment (justinjfu) [reply docs/pallas/tpu/sparse.md:564]: Done.

File: docs/pallas/tpu/sparse.md (hunk 3 above)
Comment (justinjfu) [reply docs/pallas/tpu/sparse.md:54]: Done.

File: docs/pallas/tpu/sparse.md (hunk 9 above)
Comment (justinjfu) [reply docs/pallas/tpu/sparse.md]: Done.
